### ISSUE TITLE
Flavor API should be derived at the `flutter assemble` level so it is set correctly when run from native tooling, like Android Studio or Xcode

### ISSUE THREAD

**Issue by jmagman** [flutter/flutter#155951]
Currently, the app flavor is passed into the `FLUTTER_APP_FLAVOR` based on the `--flavor` flag passed into the Flutter tool, as implemented in https://github.com/flutter/flutter/pull/134179

https://github.com/flutter/flutter/blob/9441f9d48fce1d0b425628731dd6ecab8c8b0826/packages/flutter_tools/lib/src/runner/flutter_command.dart#L1341-L1352

However, that solution doesn't work when run from native tooling, like Android Studio of Xcode, which calls only into `flutter assemble`, so `FLUTTER_APP_FLAVOR` is unset.  

The proposed community solution at https://github.com/flutter/website/issues/11178 for iOS/macOS requires `DART_DEFINES` to be individually set per Xcode scheme/config, which would be very easy to get out-of-sync, and essentially unnecessary given that `assemble` already has all the information it needs to parse which flavor should be used.

It would be better if the flavor parsing logic were pushed down into `assemble`.

**Comment by andrewkolos** [comment]
Thanks for bringing this to my attention @navaronbracke and @jmagman. Yup, I totally failed to consider users running Flutter apps from their IDE when implementing this feature (the [`appFlavor`](https://api.flutter.dev/flutter/services/appFlavor-constant.html?_gl=1*mwgipj*_ga*NDA1NDI3OTY0LjE2NjAyODQ1NzQ.*_ga_04YGWK0175*MTcyNzcyOTYyMC41MzkuMC4xNzI3NzI5NjI2LjAuMC4w) framework API).

I will see if I can implement @jmagman's suggestion of pushing this logic down to the `flutter assemble` step.

**Comment by andrewkolos** [comment]
I tried to fix this, but couldn't get it to work.

In theory, all that would need to be done is to update the `--define` parsing code within the `flutter assemble` implementation (located in [`AssembleCommand._parseDefines`](https://github.com/flutter/flutter/blob/a9daf588296d49c7ede4685b84754fb444dccb30/packages/flutter_tools/lib/src/commands/assemble.dart#L259)). Specifically, the new code would parse the flavor from `-dFlavor` (e.g. `results[kFlavor]`) and then append a ([base64-encoded](https://github.com/flutter/flutter/blob/ea208f824b5322e91bf241ccd6f19567b9a2b8dd/packages/flutter_tools/lib/src/build_info.dart#L1043)) definition for `FLUTTER_APP_FLAVOR` onto the list of dart defines.

**Comment by Sameri11** [comment]
I think, main problem for Xcode-based builds here is that build will lack `environment['FLAVOR']` If run from Xcode itself, because it is set as env variable here: https://github.com/flutter/flutter/blob/15a03a4adac970940eec3c4472037406d308d329/packages/flutter_tools/lib/src/build_info.dart#L338-L339

So, question is how to reliably obtain flavor name inside `xcode_backend.dart`, I guess?  I tried to parse it from Xcode's build configuration in a very dumb way, but @jmagman told me (in Discord's hackers-ios) that there is no enforced naming convention for  build configurations with flavors, so my implementation is very fragile and won't work in many cases.

**Comment by YaserH25** [comment]
this does affect archiving, right? cause sometimes it archive with the selected sheme and sometimes not

**Comment by navaronbracke** [comment]
> this does affect archiving, right? cause sometimes it archives with the selected sheme and sometimes not

Yes, I initially encountered this during archiving, so it is the case.

**Comment by jmagman** [comment]
I believe bundled assets https://github.com/flutter/flutter/pull/132985 have the same problem.

**Comment by vashworth** [comment]
> I believe bundled assets [#132985](https://github.com/flutter/flutter/pull/132985) have the same problem.

I confirmed this.

**Comment by vashworth** [comment]
Just keeping notes for myself:

Here's how flavors work currently (from an iOS perspective) when running through the CLI:

1. A developer can either pass [a flag with `--flavor`](https://github.com/flutter/flutter/blob/5b29db924b92eef7390afb1165c8bfe6a014249f/packages/flutter_tools/lib/src/runner/flutter_command.dart#L1448) or they can set their [default flavor in the pubspec.yaml](https://github.com/flutter/flutter/blob/5b29db924b92eef7390afb1165c8bfe6a014249f/packages/flutter_tools/lib/src/runner/flutter_command.dart#L1447).
2. The CLI then sets [`FLUTTER_APP_FLAVOR` through dart defines](https://github.com/flutter/flutter/blob/5b29db924b92eef7390afb1165c8bfe6a014249f/packages/flutter_tools/lib/src/runner/flutter_command.dart#L1449-L1463)
3. Before building the project, [`FLAVOR` is set in the build settings via the Generated.xcconfig](https://github.com/flutter/flutter/blob/5b29db924b92eef7390afb1165c8bfe6a014249f/packages/flutter_tools/lib/src/ios/xcode_build_settings.dart#L231) [from the `FLAVOR` environment variable](https://github.com/flutter/flutter/blob/5b29db924b92eef7390afb1165c8bfe6a014249f/packages/flutter_tools/lib/src/build_info.dart#L329).
4. Then when building the project, [we target the scheme using the flavor from step 1](https://github.com/flutter/flutter/blob/5b29db924b92eef7390afb1165c8bfe6a014249f/packages/flutter_tools/lib/src/ios/mac.dart#L191)
5. Then during the `build` Run Script, [`FLAVOR` is passed from the Generated.xcconfig to the environment using `-dFlavor`](https://github.com/flutter/flutter/blob/5b29db924b92eef7390afb1165c8bfe6a014249f/packages/flutter_tools/bin/xcode_backend.dart#L442)
6. Then in IosAssetBundle, `FLAVOR` is found from the [`-dFlavor` environment definitions](https://github.com/flutter/flutter/blob/5b29db924b92eef7390afb1165c8bfe6a014249f/packages/flutter_tools/lib/src/build_system/targets/ios.dart#L545)

----

Now, when ran directly through Xcode, there's a few situation.
1) `FLUTTER_APP_FLAVOR` does not get set

2) If the developer doesn't have `FLAVOR` set in their Generated.xcconfig but selects a flavor/scheme through Xcode, `FLAVOR` is null in steps 5-6 above

3) If the developer does have `FLAVOR` set in their Generated.xcconfig but selects a different flavor/scheme through Xcode, `FLAVOR` is wrong in steps 5-6 above.

**Comment by vashworth** [comment]
Okay so this is actually pretty complicated to fix... 

In Xcode world, the flavor is basically the scheme so what we really need is the scheme name is steps 5-6 of https://github.com/flutter/flutter/issues/155951#issuecomment-2634644063. However, Xcode does not pass the scheme name, and the [internet pretty much agrees it's basically impossible to get](https://stackoverflow.com/a/37592387) 😓 . You can get the `CONFIGURATION`, which may or may not have the scheme name in it. And we can't use the logic from our tooling like in step 4 because that uses the flavor passed from the flag/pubspec, which we don't have when ran through Xcode.
